perceval/backends/_slos_tree.py

```diff
@@ -0,0 +1,90 @@
+# MIT License
+#
+# Copyright (c) 2022 Quandela
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# As a special exception, the copyright holders of exqalibur library give you
+# permission to combine exqalibur with code included in the standard release of
+# Perceval under the MIT license (or modified versions of such code). You may
+# copy and distribute such a combined system following the terms of the MIT
+# license for both exqalibur and Perceval. This exception for the usage of
+# exqalibur is limited to the python bindings used by Perceval.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import exqalibur as xq
+from perceval.utils import BasicState, BSDistribution, StateVector
+from perceval.components import ACircuit
+
+from ._abstract_backends import AStrongSimulationBackend
+
+
+class SLOSTreeBackend(AStrongSimulationBackend):
+
+    def __init__(self):
+        super().__init__()
+        self._stree = xq.SLOSTree()
+
+    def set_circuit(self, circuit: ACircuit):
+        super().set_circuit(circuit)  # Computes circuit unitary as _umat
+        self._stree.set_unitary(self._umat)
+
+    def use_steiner(self, use_it: bool):
+        """
+        Use Steiner walk into the tree rather than the naive walk.
+        Steiner's method is more efficient when the number of photons is close to the number of modes.
+        """
+        self._stree.set_steiner(use_it)
+
+    def prob_amplitude(self, output_state: BasicState) -> complex:
+        istate = self._input_state
+        all_pa = self._stree.all_prob_ampli(istate)
+        return all_pa[xq.FSArray(istate.m, istate.n).find(output_state)]
```

Comment:
Do we need to create this FSArray everytime?

Reply:
Fixed